src/floor/FloorSystem.java

```diff
@@ -0,0 +1,146 @@
+			} else {
+				floors[floorToUpdate].setDownLamp(true);
+				tripTracker.addTrip(floorToUpdate, Direction.DOWN);
+				System.out.println("Floor System: Passenger requesting elevator DOWN from floor "+floorToUpdate);
+			}
+			
+			
+		}
+		
+		//if approaching floor message
+		if(m instanceof FloorTravelMessage) {
+			int elevatorToUpdate = ((FloorTravelMessage) m).getElevator();
+			int startingFloor = ((FloorTravelMessage) m).getStartingFloor();
+			int floorToArriveTo = startingFloor;
+			if (((FloorTravelMessage) m).getDirection() == FloorTravelMessage.Direction.UP) {
+				floorToArriveTo++;
+				System.out.println("Floor System: Elevator "+elevatorToUpdate+" travelling UP from floor "+startingFloor+" to floor "+floorToArriveTo);
+				
+				//update direction lamps
+				for (int i = 0; i < SimulationVars.numberOfElevators; i++) {
+					floors[i].setDirectionLamp(elevatorToUpdate, directionLampState.UP);
+				}
+			} else {
+				floorToArriveTo--;
+				System.out.println("Floor System: Elevator "+elevatorToUpdate+" travelling DOWN from floor "+startingFloor+" to floor "+floorToArriveTo);
+				
+				//update direction lamps
+				for (int i = 0; i < SimulationVars.numberOfElevators; i++) {
+					floors[i].setDirectionLamp(elevatorToUpdate, directionLampState.DOWN);
+				}
+			}
+			//arm the respective sensor to fire
+			arrivalSensors[elevatorToUpdate] = new Thread(new ArrivalSensor(elevatorToUpdate, startingFloor, floorToArriveTo));
+			arrivalSensors[elevatorToUpdate].start();
+		}
+			
+		//if arrived message
+		if (m instanceof FloorArrivalMessage) {
+				int arrivingFloor = ((FloorArrivalMessage) m).getFloor();
+				int arrivingElevator = ((FloorArrivalMessage) m).getElevator();
+				messages.FloorArrivalMessage.Direction direction = ((FloorArrivalMessage) m).getDirection();
+				System.out.println("Floor System: Elevator arrived at a floor "+arrivingFloor);
+				
+				//if a passenger is getting on going UP
+				if(floors[arrivingFloor].getUpLamp() && direction == messages.FloorArrivalMessage.Direction.UP) {
+					
+					//simulate the passenger pressing their destination
+					tripTracker.sendTripDestination(arrivingElevator, arrivingFloor, Direction.UP);
+					
+					//modify model
+					floors[arrivingFloor].setUpLamp(false);
+					System.out.println("Floor System: Passenger going UP stepped into elevator "+arrivingElevator);
+					
+				//if a passenger is getting on going DOWN
+				} else if (floors[arrivingFloor].getDownLamp() && direction == messages.FloorArrivalMessage.Direction.DOWN) {
+					
+					//simulate the passenger pressing their destination
+					tripTracker.sendTripDestination(arrivingElevator, arrivingFloor, Direction.DOWN);
+					
+					//modify model
+					floors[arrivingFloor].setDownLamp(false);
+					System.out.println("Floor System: Passenger going DOWN stepped into elevator "+arrivingElevator);
+				}
+		}
+		//keep listening forever...
+		listenAndDispatch();
```

Comment:
Does Java have tail call optimization? This probably won't matter but if we run the process long enough these stack frame's memory are going to build up

Reply:
it does not, this should be changed. https://softwareengineering.stackexchange.com/questions/272061/why-doesnt-java-have-optimization-for-tail-recursion-at-all

tl;dr things guess stack pointer locations by counting stack frames & amount of parameters in subframes